homeassistant/components/notify/pushover.py

```diff
@@ -64,13 +89,87 @@ def send_message(self, message='', **kwargs):
         if not isinstance(targets, list):
             targets = [targets]
 
+        file = {}
+
+        if file_data is not None:
+            filename = self.load_file(
+                url=file_data.get(ATTR_FILE_URL),
+                local_path=file_data.get(ATTR_FILE_PATH),
+                username=file_data.get(ATTR_FILE_USERNAME),
+                password=file_data.get(ATTR_FILE_PASSWORD),
+                auth=file_data.get(ATTR_FILE_AUTH))
+
+            if filename is not None:
+                file = (filename, open(
+                    filename, "rb"), mimetypes.guess_type(filename))
+
         for target in targets:
             if target is not None:
                 data['device'] = target
 
             try:
-                self.pushover.send_message(message, **data)
+                self.pushover.send_message(
+                    message=message, attachment=file, **data)
+
             except ValueError as val_err:
                 _LOGGER.error(str(val_err))
             except RequestError:
                 _LOGGER.exception("Could not send pushover notification")
+            finally:
+                # Remove the tempfile if one was created
+                if file_data and file_data.get(ATTR_FILE_URL) and filename:
+                    os.remove(filename)
+
+    def load_file(self, url=None, local_path=None, username=None,
+                  password=None, auth=None):
+        """Load image/document/etc from a local path or URL."""
+        try:
+            # Load the file from URL
+            if url is not None:
+                # Check whether authentication parameters are provided
+                if username is not None and password is not None:
+                    # Use digest or basic authentication
+                    if ATTR_FILE_AUTH_DIGEST == auth:
+                        auth_ = HTTPDigestAuth(username, password)
+                    else:
+                        auth_ = HTTPBasicAuth(username, password)
+                    # Load file from URL with authentication
+                    response = requests.get(
+                        url, auth=auth_, timeout=CONF_TIMEOUT)
+                else:
+                    # Load file from URL without authentication
+                    response = requests.get(url, timeout=CONF_TIMEOUT)
+
+                # Make the request and raise an error if necessary
+                response.raise_for_status()
```

Comment:
simple: check, log, return None

Reply:
That's exactly what this code does.  raise_for_status will raise an exception if there is an issue.  I then catch the error, log it, and return None.  I believe this is preferred to hard-coding in return status' and checking the response codes.  http://docs.python-requests.org/en/master/user/quickstart/